File: components/in_app_review/ios/Classes/InAppReviewPlugin.m
#import "InAppReviewPlugin.h"

@import StoreKit;
@implementation InAppReviewPlugin

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    
    FlutterMethodChannel* channel = [FlutterMethodChannel methodChannelWithName:@"dev.britannio.in_app_review" binaryMessenger:[registrar messenger]];
    
    InAppReviewPlugin* instance = [[InAppReviewPlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
    
    [self logMessage:@"handle" details:call.method];
    
    if ([call.method isEqual:@"requestReview"]) {
        [self requestReview:result];
    } else if ([call.method isEqual:@"isAvailable"]) {
        [self isAvailable:result];
    } else if ([call.method isEqual:@"openStoreListing"]) {
        [self openStoreListingWithStoreId:call.arguments result:result];
    } else {
        [self logMessage:@"method not implemented"];
        result(FlutterMethodNotImplemented);
    }
}

- (void) requestReview:(FlutterResult)result {
    if (@available(iOS 14, *)) {
        [self logMessage:@"iOS 14+"];
        UIWindowScene *scene = [self findActiveScene];
        [SKStoreReviewController requestReviewInScene:scene];
        result(nil);
    } else if (@available(iOS 10.3, *)) {
        [self logMessage:@"iOS 10.3+"];
        [SKStoreReviewController requestReview];
        result(nil);
    } else {
        result([FlutterError errorWithCode:@"unavailable"
                                   message:@"In-App Review unavailable"
                                   details:nil]);
    }
}

- (UIWindowScene *) findActiveScene  API_AVAILABLE(ios(13.0)){
    for (UIWindowScene *scene in UIApplication.sharedApplication.connectedScenes) {
        
        if (scene.activationState == UISceneActivationStateForegroundActive) {
            return scene;
        }
        
    }
    
    return nil;
}

- (void) isAvailable:(FlutterResult)result {
    if (@available(iOS 10.3, *)) {
        [self logMessage:@"available"];
        result(@YES);
    } else {
        [self logMessage:@"unavailable"];
        result(@NO);
    }
}

- (void) openStoreListingWithStoreId:(NSString *)storeId result:(FlutterResult)result {
    
    if (!storeId) {
        result([FlutterError errorWithCode:@"no-store-id"
                                   message:@"Your store id must be passed as the method channel's argument"
                                   details:nil]);
        return;
    }
    
    NSURL *url = [NSURL URLWithString:[NSString stringWithFormat:@"https://apps.apple.com/app/id%@?action=write-review", storeId]];
    
    if (!url) {
        result([FlutterError errorWithCode:@"url-construct-fail"
                                   message:@"Failed to construct url"
                                   details:nil]);
        return;
    }
        
    UIApplication *app = [UIApplication sharedApplication];
    if (@available(iOS 10.0, *)) {
        [app openURL:url options:@{} completionHandler:nil];
    } else {
        [app openURL:url];
    }
}


#pragma mark - Logging Helpers

- (void) logMessage:(NSString *) message {
    NSLog(@"InAppReviewPlugin: %@", message);
}

- (void) logMessage:(NSString *) message
            details:(NSString *) details {
    NSLog(@"InAppReviewPlugin: %@ %@", message, details);
}

@end

File: components/in_app_review/ios/in_app_review/Sources/in_app_review/InAppReviewPlugin.swift
import Flutter
import StoreKit
import UIKit

public class InAppReviewPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "dev.britannio.in_app_review",
      binaryMessenger: registrar.messenger()
    )
    let instance = InAppReviewPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    log("handle", details: call.method)

    switch call.method {
    case "requestReview":
      requestReview(result)
    case "isAvailable":
      isAvailable(result)
    case "openStoreListing":
      openStoreListing(appStoreId: call.arguments as? String, result: result)
    default:
      log("method not implemented")
      result(FlutterMethodNotImplemented)
    }
  }

  private func requestReview(_ result: @escaping FlutterResult) {
    if #available(iOS 14.0, *) {
      log("iOS 14+")
      if let scene = findActiveScene() {
        SKStoreReviewController.requestReview(in: scene)
      }
      result(nil)
    } else if #available(iOS 10.3, *) {
      log("iOS 10.3+")
      SKStoreReviewController.requestReview()
      result(nil)
    } else {
      result(
        FlutterError(
          code: "unavailable",
          message: "In-App Review unavailable",
          details: nil
        )
      )
    }
  }

  @available(iOS 13.0, *)
  private func findActiveScene() -> UIWindowScene? {
    UIApplication.shared.connectedScenes
      .compactMap { $0 as? UIWindowScene }
      .first { $0.activationState == .foregroundActive }
  }

  private func isAvailable(_ result: @escaping FlutterResult) {
    if #available(iOS 10.3, *) {
      log("available")
      result(true)
    } else {
      log("unavailable")
      result(false)
    }
  }

  private func openStoreListing(appStoreId: String?, result: @escaping FlutterResult) {
    guard let appStoreId else {
      result(
        FlutterError(
          code: "no-store-id",
          message: "Your store id must be passed as the method channel's argument",
          details: nil
        )
      )
      return
    }

    guard let url = URL(string: "https://apps.apple.com/app/id\(appStoreId)?action=write-review") else {
      result(
        FlutterError(
          code: "url-construct-fail",
          message: "Failed to construct url",
          details: nil
        )
      )
      return
    }

    UIApplication.shared.open(url)
    result(nil)
  }

  private func log(_ message: String, details: String? = nil) {
    if let details {
      NSLog("InAppReviewPlugin: \(message) \(details)")
    } else {
      NSLog("InAppReviewPlugin: \(message)")
    }
  }
}
